tests/benchmark/bm_basics.cpp

```diff
@@ -27,122 +27,163 @@ static void GetHNSWIndex(VecSimIndex *hnsw_index) {
+        VecSimIndex_AddVector(hnsw_index, (*queries)[iter++].data(), new_id++);
+    }
+    // Clean-up.
+    size_t new_index_size = VecSimIndex_IndexSize(hnsw_index);
+    for (size_t id = n_vectors; id < new_index_size; id++) {
+        VecSimIndex_DeleteVector(hnsw_index, id);
     }
 }
 
 BENCHMARK_DEFINE_F(BM_VecSimBasics, DeleteVectorHNSW)(benchmark::State &st) {
     // Remove a different vector in every execution.
+    float *blobs[1000];
+    size_t id_to_remove = 0;
+
     for (auto _ : st) {
-        size_t id = n_vectors - VecSimIndex_IndexSize(hnsw_index);
-        VecSimIndex_DeleteVector(hnsw_index, id);
+        st.PauseTiming();
+        blobs[id_to_remove] = new float[dim];
+        memcpy(blobs[id_to_remove],
+               reinterpret_cast<HNSWIndex *>(hnsw_index)
+                   ->getHNSWIndex()
+                   ->getDataByInternalId(id_to_remove),
+               dim * sizeof(float));
+        st.ResumeTiming();
+        VecSimIndex_DeleteVector(hnsw_index, id_to_remove++);
+    }
+
+    // Restore index state.
+    for (size_t i = 0; i < id_to_remove; i++) {
+        VecSimIndex_AddVector(hnsw_index, blobs[i], i);
+        delete[] blobs[i];
     }
 }
 
 BENCHMARK_DEFINE_F(BM_VecSimBasics, TopK_BF)(benchmark::State &st) {
     size_t k = st.range(0);
+    size_t iter = 0;
     for (auto _ : st) {
-        st.PauseTiming();
-        // Generate random query vector before test.
-        std::uniform_real_distribution<double> distrib(-1.0, 1.0);
-        for (size_t i = 0; i < dim; ++i) {
-            query[i] = (float)distrib(rng);
-        }
-        st.ResumeTiming();
-        VecSimIndex_TopKQuery(bf_index, query.data(), k, nullptr, BY_SCORE);
+        VecSimIndex_TopKQuery(bf_index, (*queries)[iter++].data(), k, nullptr, BY_SCORE);
     }
 }
 
 BENCHMARK_DEFINE_F(BM_VecSimBasics, TopK_HNSW)(benchmark::State &st) {
-    size_t k = st.range(0);
+    size_t ef = st.range(0);
+    size_t k = st.range(1);
     size_t correct = 0.0f;
     size_t iter = 0;
     for (auto _ : st) {
-        st.PauseTiming();
-        // Generate random query vector before test.
-        std::uniform_real_distribution<double> distrib(-1.0, 1.0);
-        for (size_t i = 0; i < dim; ++i) {
-            query[i] = (float)distrib(rng);
-        }
-        st.ResumeTiming();
-        auto hnsw_results = VecSimIndex_TopKQuery(hnsw_index, query.data(), k, nullptr, BY_SCORE);
+        auto query_params =
+            VecSimQueryParams{.hnswRuntimeParams = HNSWRuntimeParams{.efRuntime = ef}};
+        auto hnsw_results =
+            VecSimIndex_TopKQuery(hnsw_index, (*queries)[iter].data(), k, &query_params, BY_SCORE);
         st.PauseTiming();
 
         // Measure recall:
-        auto bf_results = VecSimIndex_TopKQuery(bf_index, query.data(), k, nullptr, BY_SCORE);
+        auto bf_results =
+            VecSimIndex_TopKQuery(bf_index, (*queries)[iter].data(), k, nullptr, BY_SCORE);
```

Comment:
same goes here regarding generality, maybe add a module for the iteration to not overshoot the queries array